pkg/api/authn.go

```diff
@@ -394,22 +353,65 @@ func (amw *AuthnMiddleware) TryAuthnHandlers(ctlr *Controller) mux.MiddlewareFun
 				return
 			}
 
-			//nolint: contextcheck
-			authenticated, cloneResp, cloneReq, err := amw.basicAuthn(ctlr, response, request)
-			if err != nil {
-				response.WriteHeader(http.StatusInternalServerError)
+			var authenticated bool
 
-				return
-			}
+			var cloneResp http.ResponseWriter
 
-			if authenticated && cloneResp != nil && cloneReq != nil {
-				next.ServeHTTP(cloneResp, cloneReq)
+			var cloneReq *http.Request
 
-				return
+			var err error
+
+			// try basic auth if authorization header is given
+			if !isAuthorizationHeaderEmpty(request) { //nolint: gocritic
+				//nolint: contextcheck
+				authenticated, cloneResp, cloneReq, err = amw.basicAuthn(ctlr, response, request)
+				if err != nil {
+					response.WriteHeader(http.StatusInternalServerError)
+
+					return
+				}
+
+				if authenticated && cloneResp != nil && cloneReq != nil {
+					next.ServeHTTP(cloneResp, cloneReq)
+
+					return
+				}
+			} else if hasSessionHeader(request) {
+				// try session auth
+				//nolint: contextcheck
+				authenticated, cloneResp, cloneReq, err = amw.sessionAuthn(ctlr, response, request)
+				if err != nil {
+					if errors.Is(err, zerr.ErrUserDataNotFound) {
+						ctlr.Log.Err(err).Msg("can not find user profile in DB")
+
+						authFail(response, request, ctlr.Config.HTTP.Realm, delay)
+					}
+
+					response.WriteHeader(http.StatusInternalServerError)
+
+					return
+				}
+
+				if authenticated && cloneResp != nil && cloneReq != nil {
+					next.ServeHTTP(cloneResp, cloneReq)
+
+					return
+				}
+			} else {
+				// try anonymous auth only if basic auth/session was not given
+				// we want to bypass auth for mgmt route
+				isMgmtRequested := request.RequestURI == constants.FullMgmtPrefix
+				if ctlr.Config.HTTP.AccessControl.AnonymousPolicyExists() || isMgmtRequested {
+					// do we really need this? acCtx will be non nil!
```

Comment:
Not sure what to say, I would keep the existing logic of creating the context, unless you also update all the places where it is checked for nil.